### DISCUSSION TITLE
Presentation timestamp (pts) not obeyed

### DISCUSSION THREAD

**Discussion by joshuar9** [aiortc/aiortc#824] [Q&A]
I'm developing a program that has a computer generated video. The processing rate for the video generation is faster than the realtime rate, and I want to stream the video and audio as soon as the audio and video frames are available to ensure a smooth playback on the client side with buffering. 

I'm thinking that setting the pts and framerate attributes on the AudioFrame object would allow the browser to correctly play the frames at the right time, but I'm seeing it play super fast instead of at the requested pts and framerate specified for the frames. 

The frame attributes are populated using code similar to this:
```
async def recv(self):
        self._timestamp += samples
        frame = AudioFrame(format="s16", layout="mono", samples=samples)
        for p in frame.planes:
            audioBytes = generateAudioBytes()
            p.update(audioBytes)
                    
        frame.pts = self._timestamp
        frame.sample_rate = self.sample_rate
        frame.time_base = fractions.Fraction(1, self.sample_rate)
        await asyncio.sleep(0)

        return frame
```
The video frame logic is similar. 

I've confirmed that the pts is increasing. 

I expected the pts and sample rate to be used to buffer the data so it plays at the correct playback speed, but that's not what's happening. Everything just plays really fast, and there are missing parts in the audio. lol.

The SDP for the answer has these lines, which seem to be relevant: 
```
a=rtpmap:111 opus/48000/2
a=fmtp:111 minptime=20; useinbandfec=1
a=maxptime:20
a=ptime:20
a=framerate:30
```

But that seems to have no effect on the playback speeds.

Is this a bug or am I misunderstanding what these parameters do?

**Comment by DerGenaue** [comment]
As far as I understand, `time_base` usually is 1/ the framerate of the video which in aiortc is hardcoded to be `30fps`.
Could you try converting everything to that time base instead of using the audio sample count?

I just checked the code and aiortc currently hardcodes the `time_base` to be `1/MAX_FRAMERATE` for both vpx and h264 encoders.
I assume that the encoders just completely ignore the `time_base` setting on the individual frames.
Consider adjusting your code to use this `time_base` or hack aiortc to configure the video encoders in a different way
